### PR TITLE
Bulk data

### PR DESCRIPTION
**Pull request by jmorrisnrel**
Update to implement CSV bulk data capabilities. Imports for creating/updating locations, technologies, and nodes as well as download links to download template files with existing model configurations.

UI adjustments to include bulk data links as well as a new landing page to display results/feedback from bulk data upload operations.

### CODE REVIEW

File: calliope_app/api/views/outputs.py [hunk 1]
```diff
@@ -412,4 +418,491 @@ def upload_outputs(request):
...
+
+    context = {
+                'logs':[],
+                "model": model
+              }
+
+    if (request.method == "POST") and ("myfile" in request.FILES):
+
+        myfile = request.FILES["myfile"]
+        if '.csv' in myfile.name:
```
Comment (jgu2) [inline calliope_app/api/views/outputs.py:692]: same as above.

File: calliope_app/api/views/outputs.py [hunk 2]
```diff
@@ -412,4 +418,491 @@ def upload_outputs(request):
...
+            df = pd.read_csv(myfile)
+        else:
+            context['logs'].append('File format not supported. Please use a .csv.')
+            return render(request, 'bulkresults.html', context)
+
+        if not set(['technology','location_1']).issubset(set(df.columns)):
+            context['logs'].append("Missing required columns")
```
Comment (jgu2) [inline calliope_app/api/views/outputs.py:699]: Same as above.

File: calliope_app/client/templates/configure.html [hunk 3]
```diff
@@ -9,6 +9,10 @@
 		<div class="col-1 tab" style="background-color:#dedede;z-index:-1;padding:0;z-index: 451;">
 			<button id="master-cancel" {% if not can_edit %}disabled{% endif %} type="button" class="btn btn-danger master-btn hide"><b>{% trans "Cancel" %}</b></button>
 			<button id="master-settings" {% if not can_edit %}disabled{% endif %} type="button" class="btn btn-info master-btn hide">{% trans "Settings" %}</button>
+			<div class="btn-group btn-group-vertical" style='width: 100%; height: 100%;'>
+			<button id="master-bulk-up" {% if not can_edit %}disabled{% endif %} type="button" class="btn master-btn btn-info hide" style='height: 50%;padding:0;'>{% trans "Bulk Upload" %}</button>
+			<a href="/api/bulk_download/?model_uuid={{ model.uuid }}" id="master-bulk-down" tabindex="-1" type="button" class="btn master-btn btn-info hide" style='width:101%;height:50%;padding:0;'>{% trans "Bulk Download" %}</a>
+			</div>
```
Comment (jgu2) [inline calliope_app/client/templates/configure.html:15]: use different color for upload and download on buttons?

**Review by jgu2 — COMMENTED** [review]
Fetch the upstream of public repo, rebase and solve the conflicts at the local, then push the changes, it'll solve.

File: calliope_app/api/views/outputs.py [hunk 4]
```diff
@@ -412,4 +418,491 @@ def upload_outputs(request):
...
+                            update_dict['edit']['parameter'][p.pk] = convert_units(ureg,v,p.units)
+                        except pint.errors.DimensionalityError as e:
+                            context['logs'].append(str(i)+'- Tech '+row['pretty_name']+': Column '+f+' '+str(e)+'. Parameter skipped.')
+                            continue
+                        
+            technology.update(update_dict)
+
+        return render(request, "bulkresults.html", context)
+
+    print("No File Found")
```
[the remark this answers is not in the document]
Comment (jmorrisnrel) [reply calliope_app/api/views/outputs.py]: Fixed

**Review by jgu2 — APPROVED** [review]

**Pull request merged**